File: ch07 . gateway/service/api/src/main/java/org/delivery/api/common/exception/ApiException.java
/*
package org.delivery.common.exception;

import lombok.Getter;
import org.delivery.common.error.ErrorCodeInterface;

@Getter
public class ApiException extends RuntimeException implements ApiExceptionInterface{

    private final ErrorCodeInterface errorCodeInterface;

    private final String errorDescription;

    public ApiException(ErrorCodeInterface errorCodeInterface){
        super(errorCodeInterface.getDescription());
        this.errorCodeInterface = errorCodeInterface;
        this.errorDescription = errorCodeInterface.getDescription();
    }

    public ApiException(ErrorCodeInterface errorCodeInterface, String errorDescription){
        super(errorDescription);
        this.errorCodeInterface = errorCodeInterface;
        this.errorDescription = errorDescription;
    }

    public ApiException(ErrorCodeInterface errorCodeInterface, Throwable tx){
        super(tx);
        this.errorCodeInterface = errorCodeInterface;
        this.errorDescription = errorCodeInterface.getDescription();
    }

    public ApiException(ErrorCodeInterface errorCodeInterface, Throwable tx, String errorDescription){
        super(tx);
        this.errorCodeInterface = errorCodeInterface;
        this.errorDescription = errorDescription;
    }


}
*/

File: ch07 . gateway/service/common/src/main/kotlin/org/delivery/common/exception/ApiException.kt
package org.delivery.common.exception

import org.delivery.common.error.ErrorCodeInterface

class ApiException : RuntimeException, ApiExceptionInterface {

    override val errorCodeInterface: ErrorCodeInterface
    override val errorDescription: String

    constructor(errorCodeInterface: ErrorCodeInterface) : super(errorCodeInterface.getDescription()){
        this.errorCodeInterface = errorCodeInterface
        this.errorDescription = errorCodeInterface.getDescription()
    }

    constructor(
        errorCodeInterface: ErrorCodeInterface,
        errorCodeDescription: String,
    ): super(errorCodeDescription){
        this.errorCodeInterface = errorCodeInterface
        this.errorDescription = errorCodeDescription
    }

    constructor(
        errorCodeInterface: ErrorCodeInterface,
        throwable: Throwable
    ): super(throwable){
        this.errorCodeInterface = errorCodeInterface
        this.errorDescription = errorCodeInterface.getDescription()
    }

    constructor(
        errorCodeInterface: ErrorCodeInterface,
        throwable: Throwable,
        errorCodeDescription: String
    ): super(throwable){
        this.errorCodeInterface = errorCodeInterface
        this.errorDescription = errorCodeDescription
    }
}